provider/equinix/environ.go

```diff
@@ -0,0 +1,665 @@
+		plan.Specs.Memory == nil ||
+		len(plan.Specs.Cpus) == 0 || plan.Specs.Cpus[0].Count == 0
+
+	return !isInvalid
+}
+
+func parseMemValue(v string) (uint64, error) {
+	var scaler = uint64(1)
+	if strings.HasSuffix(v, "GB") {
+		scaler = 1024
+		v = strings.TrimSuffix(v, "GB")
+	}
+
+	val, err := strconv.ParseUint(v, 10, 64)
+	return val * scaler, err
+}
+
+func (e *environ) findInstanceSpec(controller bool, allImages []*imagemetadata.ImageMetadata, instanceTypes []instances.InstanceType, ic *instances.InstanceConstraint) (*instances.InstanceSpec, error) {
+	oss, _, err := e.equnixClient.OperatingSystems.List()
+	if err != nil {
+		return nil, err
+	}
+	suitableImages := []*imagemetadata.ImageMetadata{}
+
+	for _, it := range instanceTypes {
+		for _, os := range oss {
+			if !isDistroSupported(os, ic) {
+				continue
+			}
+
+			for _, p := range os.ProvisionableOn {
+				if p == it.Name {
+					image := &imagemetadata.ImageMetadata{
+						Id:   os.Slug,
+						Arch: arch.AMD64,
+					}
+					suitableImages = append(suitableImages, image)
+				}
+			}
+		}
+	}
+
+	images := instances.ImageMetadataToImages(suitableImages)
+	return instances.FindInstanceSpec(images, ic, instanceTypes)
+}
+
+func (e *environ) finishInstanceConfig(args *environs.StartInstanceParams, spec *instances.InstanceSpec) error {
+	matchingTools, err := args.Tools.Match(tools.Filter{Arch: spec.Image.Arch})
+	if err != nil {
+		return errors.Errorf("chosen architecture %v for image %q not present in %v",
+			spec.Image.Arch, spec.Image.Id, args.Tools.Arches())
+	}
+
+	if spec.InstanceType.Deprecated {
+		logger.Infof("deprecated instance type specified: %s", spec.InstanceType.Name)
+	}
+
+	if err := args.InstanceConfig.SetTools(matchingTools); err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := instancecfg.FinishInstanceConfig(args.InstanceConfig, e.Config()); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+func (e *environ) StopInstances(ctx context.ProviderCallContext, ids ...instance.Id) error {
+	for _, id := range ids {
+		_, err := e.equnixClient.Devices.Delete(string(id), true)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (e *environ) StorageProvider(t storage.ProviderType) (storage.Provider, error) {
+	return nil, errors.NotFoundf("storage provider %q", t)
+}
+
+func (e *environ) StorageProviderTypes() ([]storage.ProviderType, error) {
+	return nil, nil
+}
+
+func waitDeviceActive(c *packngo.Client, id string) (*packngo.Device, error) {
+	err := retry.Call(retry.CallArgs{
+		Func: func() error {
+			d, _, err := c.Devices.Get(id, nil)
+			if err != nil {
+				return err
+			}
+			if d.State == "active" {
+				return nil
+			}
+			if d.State == "failed" {
+				return fmt.Errorf("device %s provisioning failed", id)
+			}
+			return nil
+		},
+		IsFatalError: func(err error) bool {
+			return common.IsCredentialNotValid(err)
+		},
+		Attempts: 180,
+		Delay:    5 * time.Second,
+		Clock:    clock.WallClock,
```

Comment:
I would recommend passing `Clock` as an argument to the function. This will allow you to substitute it with a fake clock (from the clock/test sub-package) when testing this function in your tests.